### ISSUE TITLE
Improve scale propagation in broadcasted scalar tensors.

### ISSUE THREAD

**Issue by balancap** [graphcore-research/jax-scalify#94]
We currently do not propagate any information on (constant) scalar broadcasted tensor, leading to wrong scaling estimate when converting to `ScaledArray`.

The typical example of this bug is backpropagation through `loss = -np.mean(t)`, which creates a broadcasted scalar array on the backward pass. #88 log-softmax issue is also related to this problem.

**Solution:** add metadata to array tracking in `autoscale` interpreter tracking which tensors are broadcasted scalars, and use this information when promoting to a ScaledArray.
